cpp/oneapi/dal/backend/primitives/select/select_dpc.cpp

```diff
@@ -0,0 +1,287 @@
+            const integer_t sub_group_id = sbg.get_group_id();
+            const integer_t group_id = item.get_group().get_id(0) * n_sub_groups + sub_group_id;
+
+            integer_t ind_start = group_id * elems_for_sbg;
+            integer_t ind_end =
+                sycl::min(static_cast<integer_t>((group_id + 1) * elems_for_sbg), elem_count);
+
+            integer_t group_offset = part_prefix_sum_ptr[group_id];
+
+            integer_t sum = 0;
+
+            for (integer_t i = ind_start + local_id; i < ind_end; i += local_size) {
+                const integer_t part = static_cast<integer_t>(mask_accessor[i]);
+                const integer_t boundary =
+                    group_offset + sum + exclusive_scan(sbg, part, plus<integer_t>());
+                if (part)
+                    out_ptr[boundary] = in_ptr[i];
+                sum += reduce(sbg, part, plus<integer_t>());
+            }
+        });
+    });
+
+    return event;
+}
+
+template <typename Data, typename Flag>
+select_flagged_base<Data, Flag>::select_flagged_base(const sycl::queue& queue)
+        : queue_(queue),
+          elem_count_(0) {}
+
+template <typename Data, typename Flag>
+select_flagged_base<Data, Flag>::~select_flagged_base() {
+    select_flagged_base_event_.wait_and_throw();
+}
+
+template <typename Data, typename Flag>
+void select_flagged_base<Data, Flag>::init(sycl::queue& queue, std::int64_t elem_count) {
+    ONEDAL_ASSERT(elem_count > 0);
+    ONEDAL_ASSERT(elem_count <= de::limits<integer_t>::max());
+
+    const integer_t uint_elem_count = de::integral_cast<integer_t>(elem_count);
+    if (elem_count_ != uint_elem_count) {
+        elem_count_ = uint_elem_count;
+        local_size_ = preferable_sbg_size_;
+        local_sum_count_ = de::check_mul_overflow(max_local_sum_count_, local_size_) < elem_count_
+                               ? max_local_sum_count_
+                               : (elem_count_ / local_size_) + bool(elem_count_ % local_size_);
+
+        sum_buff_size_ = local_sum_count_ + 1;
+
+        part_sum_ =
+            ndarray<integer_t, 1>::empty(queue, { sum_buff_size_ }, sycl::usm::alloc::device);
+        part_prefix_sum_ =
+            ndarray<integer_t, 1>::empty(queue, { sum_buff_size_ }, sycl::usm::alloc::device);
+        total_sum_ = ndarray<integer_t, 1>::empty(queue, { 1 }, sycl::usm::alloc::device);
+    }
+}
+
+template <typename Data, typename Flag>
+sycl::event select_flagged_base<Data, Flag>::select_flagged_base_impl(
+    const mask_t& mask_accessor,
+    const ndview<Data, 1>& in,
+    ndview<Data, 1>& out,
+    std::int64_t& selected_elem_count,
+    const event_vector& deps) {
+    ONEDAL_ASSERT(out.has_mutable_data());
+    ONEDAL_ASSERT(in.get_count() == out.get_count());
+
+    if (in.get_count() > de::limits<integer_t>::max()) {
+        /// CHange exception name
+        throw domain_error(dal::detail::error_messages::invalid_number_of_elements_to_process());
```

Comment:
Is it TODO?

Reply:
no, already changed, forget to remove comment.